src/java/org/apache/cassandra/service/snapshot/SnapshotManager.java

```diff
@@ -80,43 +81,58 @@ public Collection<TableSnapshot> getExpiringSnapshots()
         return expiringSnapshots;
     }
 
+    public Collection<TableSnapshot> getEphemeralSnapshots()
+    {
+        return ephemeralSnapshots;
+    }
+
     public synchronized void start()
     {
-        loadSnapshots();
+        if (!snapshotsLoaded)
+            loadSnapshots();
+
         resumeSnapshotCleanup();
     }
 
     public synchronized void stop() throws InterruptedException, TimeoutException
     {
         expiringSnapshots.clear();
+        ephemeralSnapshots.clear();
         if (cleanupTaskFuture != null)
         {
             cleanupTaskFuture.cancel(false);
             cleanupTaskFuture = null;
         }
+        snapshotsLoaded = false;
     }
 
     public synchronized void addSnapshot(TableSnapshot snapshot)
     {
-        // We currently only care about expiring snapshots
-        if (snapshot.isExpiring())
+        // We currently care only about expiring and ephemeral snapshots
+        if (snapshot.isExpiring() && !expiringSnapshots.contains(snapshot))
         {
-            logger.debug("Adding expiring snapshot {}", snapshot);
+            logger.debug("Adding expiring snapshot {}", snapshot.getId());
             expiringSnapshots.add(snapshot);
         }
+        // we are adding ephemeral snapshots only in case we are doing the initial loading
```

Comment:
I don't like this logic because it's inconsistent depending on the node state. We should either *always* include ephemeral, or never include ephemeral. Why do we need to load ephemeral snapshots in `SnapshotManager` in the first place? Don't we only need them to cleanup leftovers during startup? Can't we use `SnapshotLoader` to load ephemeral snapshots on `scrubDataDirectories` - no need to load on `SnapshotManager` ?

Reply:
My logic behind that was that when we put this in scrub method, we walk the disk once and then we need to walk it for the second time when we are loading snapshots in manager. I do not know if this is a big problem to scan it twice or not. I was trying to do it in such a way that we need to walk it just once. If there is 10 000 snaphots I think it starts to make a difference if we walk once or twice.

Reply:
To add to it - the problem is that if we are going to _always_ include ephemerals, you will have them during whole life of the node because they can be removed only on startup. So we will be basically only adding snapshots. Is not this a concern from memory point of view?

Reply:
Well, I mean ... we can indeed always add it and then an operator can delete these ephemerals after he sees them via nodetool, yeah. I was not taking that possibility into account. The fact that they are _always removed on startup_ does not mean that we could not remove it while a node is up, right?